hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -2427,9 +2427,17 @@ private boolean selectClauseContains(SqmFrom<?, ?> from) {
 			return true;
 		}
 		for ( SqmSelection<?> selection : selections ) {
-			if ( selection.getSelectableNode() == from ) {
+			final SqmSelectableNode<?> selectableNode = selection.getSelectableNode();
+			if ( selectableNode == from ) {
 				return true;
 			}
+			else if ( selectableNode instanceof SqmDynamicInstantiation ) {
+				for ( SqmDynamicInstantiationArgument<?> argument : ( (SqmDynamicInstantiation<?>) selectableNode ).getArguments() ) {
+					if ( argument.getSelectableNode() == from ) {
+						return true;
+					}
+				}
+			}
```

Comment:
Dynamic instantiations can be nested, so this should be a recursive check.

Reply:
Updated the PR, now `selectClauseContains` checks dynamic instantiations recursively.

I also added similar logic for `order by` expressions, which have to be expanded if they were also expanded in the `group by`, and a test for that.